### PR TITLE
Introduce Hive Dataset and New Dataloader

### PR DESCRIPTION
**Pull request by iseessel**
Summary:
This PR introduces AutoOnBoxEverstore dataset support, which accesses everstore images through a hive table. See [Important Docs here](https://fb.quip.com/587dAkUb7pyM) for more background.

Features: Should (hopefully) support all trainings/evaluations (swav+simclr tested).

1. Multi-epoch training with seeded shuffling.
2. Custom transformations and collators.
3. Dataloader state checkpointing.
4. Support for distributed training with DDP.

Limitations
1. Multiple data sources not supported if using hive.
2. Must have label inside hive table.
3. Must specify DATA_LIMIT, will not read entire table.
4. More?

Next Steps
1. Run full 8node, 8gpu, 100epoch SwaV train/eval with imagenet to test accuracies/speed of trainings.
2. Create unit and integration tests.
3. Cleanup code: Comments, types, asserts, etc.,
4. Create followup diff for "multiple_phases_per_epoch" support (important for large datasets).
5. Post about new feature in workplace.

Differential Revision: D28080328

### CODE REVIEW

File: vissl/data/ssl_transforms/tensor_to_img_pil.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    def __call__(self, sample):
+        # Convert tensors to image pil for subsequent transforms.
+        for i, tensor in enumerate(sample["data"]):
+            sample["data"][i] = self._tensor_to_img_pil(tensor)
+
+        sample = self._reformat_sample(sample)
+
+        return sample
+
+    def _tensor_to_img_pil(self, tensor):
```
Comment (QuentinDuval) [inline vissl/data/ssl_transforms/tensor_to_img_pil.py:26]: There is a torchvision transformation that might be used here (to move from a tensor to a PIL image): https://pytorch.org/vision/stable/_modules/torchvision/transforms/transforms.html#ToPILImage

File: vissl/trainer/train_steps/standard_train_step.py [hunk 2]
```diff
@@ -118,6 +118,7 @@ def standard_train_step(task):
     with PerfTimer("read_sample", perf_stats):
         sample = next(task.data_iterator)
 
+    sample = task.post_process_batch(sample)
```
Comment (QuentinDuval) [inline vissl/trainer/train_steps/standard_train_step.py:121]: 👍

File: vissl/trainer/train_task.py [hunk 3]
```diff
@@ -279,43 +277,70 @@ def _build_phases(self):
...
                 device=self.device,
                 sampler_seed=sampler_seed,
+                split=split.lower(),
             )
             for split in self.available_splits
         }
 
         return loaders
 
+    def build_dataloader_iterator(self, phase_type: str):
```
Comment (QuentinDuval) [inline vissl/trainer/train_task.py:324]: 👍

File: vissl/trainer/train_task.py [hunk 4]
```diff
@@ -696,7 +703,37 @@ def prepare(self, pin_memory: bool = False):
...
+            pin_memory=pin_memory, current_phase_id=current_phase_id
+        )
+        task = self._update_classy_state(vissl_state_dict)
+
+        return task
+
+    def set_epoch(self, phase_type: str, epoch: int, start_iter: int):
+        if hasattr(self.dataloaders[phase_type], "sampler"):
+            sampler = self.dataloaders[phase_type].sampler
+            # (Re-)Shuffle data: set epoch of distributed (or fairstore) sampler.
```
Comment (QuentinDuval) [inline vissl/trainer/train_task.py:719]: nit: the following lines could be factorized with below

```
if hasattr(sampler, "set_epoch"):
    sampler.set_epoch(epoch)
# Resume from the iteration if valid
if hasattr(sampler, "set_start_iter"):
    sampler.set_start_iter(start_iter)
```

Another idea would be to introduce a "tagging interface": instead of checking the presence of method, do a `isinstance` check of an interface, that contains the functions you want. The main advantage is that the name of the interface might bring some more additional information.

File: vissl/data/ssl_dataset.py [hunk 5]
```diff
@@ -33,7 +36,7 @@ def _convert_lbl_to_long(lbl):
     return out_lbl
 
 
-class GenericSSLDataset(Dataset):
+class GenericSSLDataset(VisslDatasetBase):
```
Comment (QuentinDuval) [inline vissl/data/ssl_dataset.py:39]: 👍 👍

File: vissl/data/ssl_transforms/__init__.py [hunk 6]
```diff
@@ -4,185 +4,9 @@
...
-        transform = SSLTransformsWrapper.from_config(transform_config)
-        output_transforms.append(transform)
-    return pth_transforms.Compose(output_transforms)
-
-
 FILE_ROOT = Path(__file__).parent
-import_all_modules(FILE_ROOT, "vissl.data.ssl_transforms")
 
-__all__ = ["SSLTransformsWrapper", "get_transform"]
+import_all_modules(FILE_ROOT, "vissl.data.ssl_transforms")
```
Comment (QuentinDuval) [inline vissl/data/ssl_transforms/__init__.py:12]: 👍 👍

File: vissl/data/vissl_dataset_base.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
+from typing import Callable, Dict, Set
+
+import torch
+from torch.utils.data import Dataset
+from vissl.config import AttrDict
+
+
+class VisslDatasetBase(Dataset):
```
Comment (QuentinDuval) [inline vissl/data/vissl_dataset_base.py:8]: nit: consider using `abc` to define the abstract methods

```
class Base(abc.ABC):
   @abc.abstractmethod
   def f(self):
      ...
```

This makes the class and sub class not instantiable if they do not define the missing functions

File: vissl/data/ssl_transforms/tensor_to_img_pil.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+from io import BytesIO
+
+from classy_vision.dataset.transforms import register_transform
+from classy_vision.dataset.transforms.classy_transform import ClassyTransform
+from PIL import Image
+
+
+@register_transform("TensorToImgPil")
+class TensorToImgPil(ClassyTransform):
```
[the remark this answers is not in the document]
Comment (iseessel) [reply vissl/data/ssl_transforms/tensor_to_img_pil.py:11]: Yeah -- not really. It should probably be moved to internal.

I refrained from doing this because it would then require some refactoring of the code here: https://github.com/facebookresearch/vissl/pull/317/files#diff-82aa56b0a134fa104a6d729561ff3aaf76e2b69c735037d7c8858480dafc5622R140 to be because this class would no longer be importable.

File: vissl/data/vissl_dataset_base.py (hunk 7 above)
Comment (iseessel) [reply vissl/data/vissl_dataset_base.py:8]: Yeah, I wasn't sure whether or not we needed the datasets to inherit from the torch Dataset, I'll test.

File: vissl/trainer/train_task.py (hunk 4 above)
Comment (iseessel) [reply vissl/trainer/train_task.py:724]: Yep indeed, good catch (this code is existing, just moved).

IMO, I am not loving this `has_attr` pattern in general. I would prefer an object oriented interface w/ default no-op methods that can be overwritten.

File: vissl/data/ssl_transforms/tensor_to_img_pil.py (hunk 1 above)
Comment (iseessel) [reply vissl/data/ssl_transforms/tensor_to_img_pil.py:26]: So I tried this, it doesn't work unfortunately. The tensor here is flattened, and torchvision expects a C x W x H tensor.

File: vissl/data/ssl_transforms/tensor_to_img_pil.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (iseessel) [reply vissl/data/ssl_transforms/tensor_to_img_pil.py:14]: I move to internal in subsequent PR.
